### PR TITLE
Added precheck for scheduler

### PR DESCRIPTION
**Pull request by evalsocket**
# TL;DR
Flytescheduler has a dependency on flyteadmin. flytescheduler will fail if flyteadmin is not available....it will use as an init container for scheduler deployment 

- Added a new command in scheduler to check flyteadmin health 

Pre request: https://github.com/flyteorg/flyte/pull/1374

Test With Sandbox :
```bash
$ flytectl sandbox start
$ Clone & checkout the PR https://github.com/flyteorg/flyte/pull/1374
$ cd charts/flyte
$ changes flytescheduler image to `evalsocket/flytescheduler`  and tag to latest. 
$ Set kubeconfig for sandbox
$ helm uninstall flyte -n flyte
$  helm install -n flyte -f values-sandbox.yaml --create-namespace flyte . 
```

Test On Local :
```
$ flytectl sandbox start
$ Clone & checkout the PR https://github.com/flyteorg/flyteadmin/pull/254
$ go run cmd/scheduler/main.go precheck --config flyteadmin_config.yaml 
$ kubectl port-forward deployment/flyteadmin 8080:8080 -n flyte 
$ add admin config in flyteadmin_config.yaml
$ go run cmd/scheduler/main.go precheck --config flyteadmin_config.yaml
```

Not tested with auth & SSL.(https://github.com/flyteorg/flyte/issues/1471)

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
_Remove the '*fixes*' keyword if there will be multiple PRs to fix the linked issue_

fixes https://github.com/flyteorg/flyte/issues/<number>

## Follow-up issue
_NA_
OR
_https://github.com/flyteorg/flyte/issues/<number>_

### CODE REVIEW

File: cmd/scheduler/entrypoints/precheck.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+var preCheckRunCmd = &cobra.Command{
+	Use:   "precheck",
+	Short: "This command will check pre requirement for scheduler",
+	RunE: func(cmd *cobra.Command, args []string) error {
+		ctx := context.Background()
+		err := retry.Do(
+			func() error {
+				config := adminClient.GetConfig(ctx)
+				host := strings.Split(config.Endpoint.Host, ":")
+				response, err := http.Get(fmt.Sprintf("http://%s/healthcheck", host))
```
Comment (pmahindrakar-oss) [inline cmd/scheduler/entrypoints/precheck.go:26]: Isn't there a grpc call to do the health check

File: flyteadmin_config.yaml [hunk 2]
```diff
@@ -206,3 +206,5 @@ qualityOfService:
     # by default production has an UNDEFINED tier when it is omitted from the configuration
 namespace_mapping:
    template: "{{ project }}-{{ domain }}" # Default namespace mapping template.
+admin:
+  endpoint: "dns:///localhost:30081"
```
Comment (pmahindrakar-oss) [inline flyteadmin_config.yaml:210]: Why is this needed? Also when running in local mode this  portwont work since we dont bring up ingress and instead would require 8088 /8089

File: cmd/scheduler/entrypoints/precheck.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+				if err != nil {
+					return err
+				}
+				if response.StatusCode == 200 {
+					return nil
+				}
+				return errors.New("something goes wrong")
```
[the remark this answers is not in the document]
Comment (evalsocket) [reply cmd/scheduler/entrypoints/precheck.go]: Now we don't need

File: cmd/scheduler/entrypoints/precheck.go (hunk 1 above)
Comment (evalsocket) [reply cmd/scheduler/entrypoints/precheck.go]: Yes added grpc call

File: flyteadmin_config.yaml (hunk 2 above)
Comment (evalsocket) [reply flyteadmin_config.yaml]: yes removed it for now

File: cmd/scheduler/entrypoints/precheck.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+						return fmt.Errorf("timeout: health rpc did not complete within %v", timeout)
+					}
+					return err
+				}
+				if resp.GetStatus() != healthpb.HealthCheckResponse_SERVING {
+					return errors.New("Health check failed")
+				}
+				return nil
+			},
+			retry.Delay(10*time.Second),
```
Comment (EngHabu) [inline cmd/scheduler/entrypoints/precheck.go:71]: Can this exponentially backoff?

File: cmd/scheduler/entrypoints/precheck.go [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+				}
+				rpcCtx := metadata.NewOutgoingContext(ctx, metadata.MD{})
+				resp, err := healthpb.NewHealthClient(conn).Check(rpcCtx,
+					&healthpb.HealthCheckRequest{
+						Service: "",
+					})
+				if err != nil {
+					if stat, ok := status.FromError(err); ok && stat.Code() == codes.Unimplemented {
+						logger.Printf(ctx, "terror: this server does not implement the grpc health protocol (grpc.health.v1.Health): %s", stat.Message())
+						return fmt.Errorf("terror: this server does not implement the grpc health protocol (grpc.health.v1.Health): %s", stat.Message())
```
Comment (EngHabu) [inline cmd/scheduler/entrypoints/precheck.go:59]: This error shouldn't be retried... will it be?

File: cmd/scheduler/entrypoints/precheck.go [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+					return fmt.Errorf("error: failed to connect service at %q: %+v", config.Endpoint.String(), err)
+				}
+				rpcCtx := metadata.NewOutgoingContext(ctx, metadata.MD{})
+				resp, err := healthpb.NewHealthClient(conn).Check(rpcCtx,
+					&healthpb.HealthCheckRequest{
+						Service: "",
+					})
+				if err != nil {
+					if stat, ok := status.FromError(err); ok && stat.Code() == codes.Unimplemented {
+						logger.Printf(ctx, "terror: this server does not implement the grpc health protocol (grpc.health.v1.Health): %s", stat.Message())
```
Comment (pmahindrakar-oss) [inline cmd/scheduler/entrypoints/precheck.go:58]: Can we just return the err and have the main routine deal with printing the error similar to how we do in flytectl.
Similarly in L45 and L48, L61

File: cmd/scheduler/entrypoints/precheck.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+						return fmt.Errorf("terror: this server does not implement the grpc health protocol (grpc.health.v1.Health): %s", stat.Message())
+					} else if stat, ok := status.FromError(err); ok && stat.Code() == codes.DeadlineExceeded {
+						logger.Printf(ctx, "timeout: health rpc did not complete within %v", timeout)
+						return fmt.Errorf("timeout: health rpc did not complete within %v", timeout)
+					}
+					return err
+				}
+				if resp.GetStatus() != healthpb.HealthCheckResponse_SERVING {
+					return errors.New("Health check failed")
```
Comment (pmahindrakar-oss) [inline cmd/scheduler/entrypoints/precheck.go:67]: due to add the status aswell in the log statement

File: cmd/scheduler/entrypoints/precheck.go [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+				}
+				return nil
+			},
+			retry.Delay(10*time.Second),
+			retry.Attempts(100),
```
Comment (pmahindrakar-oss) [inline cmd/scheduler/entrypoints/precheck.go:72]: constants and have them defined at the top

File: cmd/scheduler/entrypoints/precheck.go [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+		opts := []grpc.DialOption{
+			grpc.WithUserAgent("grpc_health_probe"),
+			grpc.WithBlock(),
+			grpc.WithInsecure(),
+		}
+
+		err := retry.Do(
+			func() error {
+				ctx := context.Background()
+				config := adminClient.GetConfig(ctx)
```
Comment (pmahindrakar-oss) [inline cmd/scheduler/entrypoints/precheck.go:39]: What happens when admin is configured with auth and ssl

File: cmd/scheduler/entrypoints/precheck.go (hunk 7 above)
Comment (evalsocket) [reply cmd/scheduler/entrypoints/precheck.go:66]: we will print err in main scope

File: cmd/scheduler/entrypoints/precheck.go (hunk 9 above)
Comment (evalsocket) [reply cmd/scheduler/entrypoints/precheck.go:39]: Currently auth is not tested and SSL will be handle by the ingress

File: cmd/scheduler/entrypoints/precheck.go (hunk 6 above)
Comment (evalsocket) [reply cmd/scheduler/entrypoints/precheck.go]: the log will help or else it will only print the error after retry exhausted

**Review by pmahindrakar-oss — APPROVED** [review]

File: cmd/scheduler/entrypoints/precheck.go [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+	"github.com/spf13/cobra"
+)
+
+const (
+	timeout            = 30 * time.Second
+	timeoutError       = "timeout: failed to connect service %q within %v"
+	connectionError    = "error: failed to connect service at %q: %+v"
+	deadlineError      = "timeout: health rpc did not complete within %v"
+	healthCheckError   = "Health check failed"
```
Comment (pmahindrakar-oss) [inline cmd/scheduler/entrypoints/precheck.go:27]: add `due to %v status`

**Review by pmahindrakar-oss — APPROVED** [review]
